File: internal/types/evaluation.go
package types

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"

	"github.com/yanyiwu/gojieba"
)

// Jieba is a global instance of Chinese text segmentation tool
var Jieba *gojieba.Jieba = newJieba()

func newJieba() *gojieba.Jieba {
	dictDir := os.Getenv("JIEBA_DICT_DIR")
	if dictDir == "" {
		return gojieba.NewJieba()
	}

	return gojieba.NewJieba(
		filepath.Join(dictDir, "jieba.dict.utf8"),
		filepath.Join(dictDir, "hmm_model.utf8"),
		filepath.Join(dictDir, "user.dict.utf8"),
		filepath.Join(dictDir, "idf.utf8"),
		filepath.Join(dictDir, "stop_words.utf8"),
	)
}

// EvaluationStatue represents the status of an evaluation task
type EvaluationStatue int

const (
	EvaluationStatuePending EvaluationStatue = iota // Task is waiting to start
	EvaluationStatueRunning                         // Task is in progress
	EvaluationStatueSuccess                         // Task completed successfully
	EvaluationStatueFailed                          // Task failed
)

// EvaluationTask contains information about an evaluation task
type EvaluationTask struct {
	ID        string `json:"id"`         // Unique task ID
	TenantID  uint64 `json:"tenant_id"`  // Tenant/Organization ID
	DatasetID string `json:"dataset_id"` // Dataset ID for evaluation

	StartTime time.Time        `json:"start_time"`        // Task start time
	Status    EvaluationStatue `json:"status"`            // Current task status
	ErrMsg    string           `json:"err_msg,omitempty"` // Error message if failed

	Total    int `json:"total,omitempty"`    // Total items to evaluate
	Finished int `json:"finished,omitempty"` // Completed items count
}

// EvaluationDetail contains detailed evaluation information
type EvaluationDetail struct {
	Task   *EvaluationTask      `json:"task"`             // Evaluation task info
	Params *ChatManage          `json:"params"`           // Evaluation parameters
	Config *EvaluationRunConfig `json:"config,omitempty"` // Effective reproducibility configuration
	Metric *MetricResult        `json:"metric,omitempty"` // Evaluation metrics
	// Result is the stage-one, in-memory four-dimension observation result.
	// Metric remains available for backwards compatibility.
	Result *EvaluationRunResult `json:"result,omitempty"`
}

// Evaluation result schema and status constants are strings because they are
// part of the public JSON contract. They are deliberately separate from the
// legacy integer EvaluationStatue used by EvaluationTask.
const EvaluationResultSchemaVersion = "evaluation-run/v1"

type EvaluationRunStatus string

const (
	EvaluationRunStatusPending EvaluationRunStatus = "pending"
	EvaluationRunStatusRunning EvaluationRunStatus = "running"
	EvaluationRunStatusSuccess EvaluationRunStatus = "success"
	EvaluationRunStatusPartial EvaluationRunStatus = "partial"
	EvaluationRunStatusFailed  EvaluationRunStatus = "failed"
)

type EvaluationPhase string

const (
	EvaluationPhasePreparation EvaluationPhase = "preparation"
	EvaluationPhaseEvaluation  EvaluationPhase = "evaluation"
	EvaluationPhaseCleanup     EvaluationPhase = "cleanup"
)

type EvaluationUsageSource string

const (
	EvaluationUsageSourceProviderReported EvaluationUsageSource = "provider_reported"
	EvaluationUsageSourceUnavailable      EvaluationUsageSource = "unavailable"
)

type EvaluationUsageStatus string

const (
	EvaluationUsageStatusComplete      EvaluationUsageStatus = "complete"
	EvaluationUsageStatusPartial       EvaluationUsageStatus = "partial"
	EvaluationUsageStatusUnavailable   EvaluationUsageStatus = "unavailable"
	EvaluationUsageStatusNotApplicable EvaluationUsageStatus = "not_applicable"
)

type EvaluationCostStatus string

const (
	EvaluationCostStatusComplete      EvaluationCostStatus = "complete"
	EvaluationCostStatusPartial       EvaluationCostStatus = "partial"
	EvaluationCostStatusUnavailable   EvaluationCostStatus = "unavailable"
	EvaluationCostStatusNotApplicable EvaluationCostStatus = "not_applicable"
)

type EvaluationModelType string

const (
	EvaluationModelTypeChat      EvaluationModelType = "chat"
	EvaluationModelTypeEmbedding EvaluationModelType = "embedding"
	EvaluationModelTypeRerank    EvaluationModelType = "rerank"
)

type EvaluationModelOperation string

const (
	EvaluationOperationChat       EvaluationModelOperation = "chat"
	EvaluationOperationEmbed      EvaluationModelOperation = "embed"
	EvaluationOperationBatchEmbed EvaluationModelOperation = "batch_embed"
	EvaluationOperationRerank     EvaluationModelOperation = "rerank"
)

// EvaluationRunResult combines the existing quality metrics with application-
// side model usage and timing observations. Persistence stores this response
// contract as a versioned snapshot rather than recalculating historical runs.
type EvaluationRunResult struct {
	SchemaVersion string                     `json:"schema_version"`
	Run           EvaluationRunMetadata      `json:"run"`
	Retrieval     *EvaluationRetrievalResult `json:"retrieval"`
	Answer        *EvaluationAnswerResult    `json:"answer"`
	Usage         EvaluationUsageResult      `json:"usage"`
	Cost          EvaluationCostResult       `json:"cost"`
	Timing        EvaluationTimingResult     `json:"timing"`
	Cases         []EvaluationCaseResult     `json:"cases"`
	Warnings      []EvaluationWarning        `json:"warnings"`
}

type EvaluationRunMetadata struct {
	RunID          string              `json:"run_id"`
	TenantID       uint64              `json:"tenant_id"`
	DatasetID      string              `json:"dataset_id"`
	StartedAt      time.Time           `json:"started_at"`
	CompletedAt    *time.Time          `json:"completed_at"`
	Status         EvaluationRunStatus `json:"status"`
	PricingVersion string              `json:"pricing_version,omitempty"`
}

type EvaluationRetrievalResult struct {
	Precision float64 `json:"precision"`
	Recall    float64 `json:"recall"`
	NDCG3     float64 `json:"ndcg3"`
	NDCG10    float64 `json:"ndcg10"`
	MRR       float64 `json:"mrr"`
	MAP       float64 `json:"map"`
}

type EvaluationAnswerResult struct {
	BLEU1  float64 `json:"bleu1"`
	BLEU2  float64 `json:"bleu2"`
	BLEU4  float64 `json:"bleu4"`
	ROUGE1 float64 `json:"rouge1"`
	ROUGE2 float64 `json:"rouge2"`
	ROUGEL float64 `json:"rougel"`
}

type EvaluationCallCounts struct {
	Total     int `json:"total"`
	Succeeded int `json:"succeeded"`
	Failed    int `json:"failed"`
	Items     int `json:"items"`
}

type EvaluationTokenTotals struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
	CachedTokens     int `json:"cached_tokens"`
	CacheReadTokens  int `json:"cache_read_tokens"`
	CacheWriteTokens int `json:"cache_write_tokens"`
	CacheMissTokens  int `json:"cache_miss_tokens"`
}

type EvaluationUsageResult struct {
	Status               EvaluationUsageStatus  `json:"status"`
	Calls                EvaluationCallCounts   `json:"calls"`
	Tokens               EvaluationTokenTotals  `json:"tokens"`
	CacheStatus          string                 `json:"cache_status"`
	ReportedCallCount    int                    `json:"reported_call_count"`
	UnavailableCallCount int                    `json:"unavailable_call_count"`
	ByModel              []EvaluationModelUsage `json:"by_model"`
	ByPhase              []EvaluationPhaseUsage `json:"by_phase"`
}

type EvaluationModelUsage struct {
	ModelType   EvaluationModelType      `json:"model_type"`
	ModelID     string                   `json:"model_id"`
	ModelName   string                   `json:"model_name"`
	Operation   EvaluationModelOperation `json:"operation"`
	UsageSource EvaluationUsageSource    `json:"usage_source"`
	Calls       EvaluationCallCounts     `json:"calls"`
	Tokens      EvaluationTokenTotals    `json:"tokens"`
	DurationMS  int64                    `json:"duration_ms"`
}

type EvaluationPhaseUsage struct {
	Phase       EvaluationPhase       `json:"phase"`
	UsageSource EvaluationUsageSource `json:"usage_source"`
	Calls       EvaluationCallCounts  `json:"calls"`
	Tokens      EvaluationTokenTotals `json:"tokens"`
	DurationMS  int64                 `json:"duration_ms"`
}

type EvaluationCostResult struct {
	Status         EvaluationCostStatus `json:"status"`
	Source         string               `json:"source"`
	Currency       string               `json:"currency,omitempty"`
	Amount         *float64             `json:"amount"`
	PricingVersion string               `json:"pricing_version,omitempty"`
	Warnings       []EvaluationWarning  `json:"warnings"`
}

type EvaluationTimingResult struct {
	TotalWallTimeMS       int64 `json:"total_wall_time_ms"`
	PreparationMS         int64 `json:"preparation_ms"`
	EvaluationMS          int64 `json:"evaluation_ms"`
	CleanupMS             int64 `json:"cleanup_ms"`
	CaseCount             int   `json:"case_count"`
	CaseAverageMS         int64 `json:"case_avg_ms"`
	CaseMinimumMS         int64 `json:"case_min_ms"`
	CaseMaximumMS         int64 `json:"case_max_ms"`
	CaseP50MS             int64 `json:"case_p50_ms"`
	CaseP95MS             int64 `json:"case_p95_ms"`
	ModelCallCumulativeMS int64 `json:"model_call_cumulative_ms"`
}

type EvaluationCaseResult struct {
	CaseID      string                 `json:"case_id"`
	Status      EvaluationRunStatus    `json:"status"`
	StartedAt   time.Time              `json:"started_at"`
	CompletedAt *time.Time             `json:"completed_at"`
	DurationMS  int64                  `json:"duration_ms"`
	Usage       EvaluationUsageResult  `json:"usage"`
	Evidence    EvaluationCaseEvidence `json:"evidence"`
	Warnings    []EvaluationWarning    `json:"warnings"`
}

// EvaluationCaseEvidence retains non-text inputs and outputs needed to audit one case score.
type EvaluationCaseEvidence struct {
	QID                        int           `json:"qid"`
	QuestionFingerprint        string        `json:"question_fingerprint"`
	ReferenceAnswerFingerprint string        `json:"reference_answer_fingerprint"`
	GeneratedAnswerFingerprint string        `json:"generated_answer_fingerprint"`
	GroundTruthPIDs            []int         `json:"ground_truth_pids"`
	SearchPIDs                 []int         `json:"search_pids"`
	RerankPIDs                 []int         `json:"rerank_pids"`
	MetricInputPIDs            []int         `json:"metric_input_pids"`
	UnmappedResultCount        int           `json:"unmapped_result_count"`
	Metrics                    *MetricResult `json:"metrics,omitempty"`
	FailureStage               string        `json:"failure_stage,omitempty"`
}

type EvaluationWarning struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// String returns JSON representation of EvaluationTask
func (e *EvaluationTask) String() string {
	b, _ := json.Marshal(e)
	return string(b)
}

// MetricInput contains input data for metric calculation
type MetricInput struct {
	RetrievalGT  [][]int // Ground truth for retrieval
	RetrievalIDs []int   // Retrieved IDs

	GeneratedTexts string // Generated text for evaluation
	GeneratedGT    string // Ground truth text for comparison
}

// MetricResult contains evaluation metrics
type MetricResult struct {
	RetrievalMetrics  RetrievalMetrics  `json:"retrieval_metrics"`  // Retrieval performance metrics
	GenerationMetrics GenerationMetrics `json:"generation_metrics"` // Text generation quality metrics
}

// RetrievalMetrics contains metrics for retrieval evaluation
type RetrievalMetrics struct {
	Precision float64 `json:"precision"` // Precision score
	Recall    float64 `json:"recall"`    // Recall score

	NDCG3  float64 `json:"ndcg3"`  // Normalized Discounted Cumulative Gain at 3
	NDCG10 float64 `json:"ndcg10"` // Normalized Discounted Cumulative Gain at 10
	MRR    float64 `json:"mrr"`    // Mean Reciprocal Rank
	MAP    float64 `json:"map"`    // Mean Average Precision
}

// GenerationMetrics contains metrics for text generation evaluation
type GenerationMetrics struct {
	BLEU1 float64 `json:"bleu1"` // BLEU-1 score
	BLEU2 float64 `json:"bleu2"` // BLEU-2 score
	BLEU4 float64 `json:"bleu4"` // BLEU-4 score

	ROUGE1 float64 `json:"rouge1"` // ROUGE-1 score
	ROUGE2 float64 `json:"rouge2"` // ROUGE-2 score
	ROUGEL float64 `json:"rougel"` // ROUGE-L score
}

// EvalState represents different stages of evaluation process
type EvalState int

const (
	StateBegin             EvalState = iota // Evaluation started
	StateAfterQaPairs                       // After loading QA pairs
	StateAfterDataset                       // After processing dataset
	StateAfterEmbedding                     // After generating embeddings
	StateAfterVectorSearch                  // After vector search
	StateAfterRerank                        // After reranking
	StateAfterComplete                      // After completion
	StateEnd                                // Evaluation ended
)

File: frontend/src/api/evaluation.ts
import { get, post } from '@/utils/request'

export type EvaluationRunStatus = 'pending' | 'running' | 'success' | 'partial' | 'failed'

export interface EvaluationWarning {
  code: string
  message: string
}

export interface EvaluationDatasetFile {
  name: string
  fingerprint: string
  size: number
}

export interface EvaluationDatasetDescriptor {
  id: string
  version: string
  content_fingerprint: string
  files?: EvaluationDatasetFile[]
  query_count: number
  corpus_count: number
  case_count: number
  ingestion_mode: string
}

export interface EvaluationModelConfig {
  id: string
  name: string
  display_name?: string
  type: string
  source: string
  provider?: string
  interface_type?: string
  embedding_dimension?: number
  max_concurrency?: number
  endpoint_fingerprint?: string
  parameters_fingerprint: string
  updated_at: string
}

export interface EvaluationModelConfigSet {
  embedding: EvaluationModelConfig
  chat: EvaluationModelConfig
  rerank?: EvaluationModelConfig | null
}

export interface EvaluationReproducibility {
  status: string
  warnings: EvaluationWarning[]
}

export interface EvaluationRunConfig {
  schema_version: string
  dataset: EvaluationDatasetDescriptor
  source_knowledge_base_id?: string
  models: EvaluationModelConfigSet
  chunking: {
    applied: boolean
    source_unit: string
    config: Record<string, unknown>
  }
  retrieval: Record<string, number>
  generation: Record<string, unknown>
  indexing: {
    vector_enabled: boolean
    keyword_enabled: boolean
    vector_store_id?: string
  }
  runtime: {
    case_concurrency: number
    metric_version: string
    result_version: string
    application_version?: string
    commit_sha: string
    vcs_modified: boolean
    commit_available: boolean
  }
  reproducibility: EvaluationReproducibility
  config_hash: string
}

export interface EvaluationRetrievalResult {
  precision: number
  recall: number
  ndcg3: number
  ndcg10: number
  mrr: number
  map: number
}

export interface EvaluationAnswerResult {
  bleu1: number
  bleu2: number
  bleu4: number
  rouge1: number
  rouge2: number
  rougel: number
}

export interface EvaluationMetrics {
  retrieval_metrics: EvaluationRetrievalResult
  generation_metrics: EvaluationAnswerResult
}

export interface EvaluationCallCounts {
  total: number
  succeeded: number
  failed: number
  items: number
}

export interface EvaluationTokenTotals {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  cached_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  cache_miss_tokens: number
}

export interface EvaluationUsageResult {
  status: string
  calls: EvaluationCallCounts
  tokens: EvaluationTokenTotals
  cache_status: string
  reported_call_count: number
  unavailable_call_count: number
  by_model: Array<Record<string, unknown>>
  by_phase: Array<Record<string, unknown>>
}

export interface EvaluationCostResult {
  status: string
  source: string
  currency?: string
  amount: number | null
  pricing_version?: string
  warnings: EvaluationWarning[]
}

export interface EvaluationTimingResult {
  total_wall_time_ms: number
  preparation_ms: number
  evaluation_ms: number
  cleanup_ms: number
  case_count: number
  case_avg_ms: number
  case_min_ms: number
  case_max_ms: number
  case_p50_ms: number
  case_p95_ms: number
  model_call_cumulative_ms: number
}

export interface EvaluationCaseEvidence {
  qid: number
  question_fingerprint: string
  reference_answer_fingerprint: string
  generated_answer_fingerprint: string
  ground_truth_pids: number[]
  search_pids: number[]
  rerank_pids: number[]
  metric_input_pids: number[]
  unmapped_result_count: number
  metrics?: EvaluationMetrics
  failure_stage?: string
}

export interface EvaluationCaseResult {
  case_id: string
  status: EvaluationRunStatus
  started_at: string
  completed_at: string | null
  duration_ms: number
  usage: EvaluationUsageResult
  evidence: EvaluationCaseEvidence
  warnings: EvaluationWarning[]
}

export interface EvaluationRunProgress {
  total: number
  finished: number
  cases: {
    total: number
    pending: number
    running: number
    success: number
    partial: number
    failed: number
  }
}

export interface EvaluationRunSummary {
  run_id: string
  status: EvaluationRunStatus
  dataset: EvaluationDatasetDescriptor
  source_knowledge_base_id?: string
  config_hash: string
  config_schema_version: string
  metric_version: string
  result_version: string
  models: EvaluationModelConfigSet
  reproducibility: EvaluationReproducibility
  progress: EvaluationRunProgress
  retrieval: EvaluationRetrievalResult | null
  answer: EvaluationAnswerResult | null
  usage: EvaluationUsageResult
  cost: EvaluationCostResult
  timing: EvaluationTimingResult
  warnings: EvaluationWarning[]
  error_message?: string
  started_at: string
  completed_at: string | null
  created_at: string
  updated_at: string
}

export interface EvaluationRunPage {
  items: EvaluationRunSummary[]
  total: number
  page: number
  page_size: number
}

export interface EvaluationRunOverview {
  summary: EvaluationRunSummary
  config: EvaluationRunConfig | null
  metric?: EvaluationMetrics
}

export interface EvaluationCasePage {
  items: EvaluationCaseResult[]
  total: number
  page: number
  page_size: number
}

export interface EvaluationComparisonCompatibility {
  comparable: boolean
  reasons: string[]
  warnings: string[]
}

export interface EvaluationValueDelta {
  baseline: number | null
  value: number | null
  absolute: number | null
  percent: number | null
}

export type EvaluationQualityDeltas = Record<
  'precision' | 'recall' | 'ndcg3' | 'ndcg10' | 'mrr' | 'map' | 'bleu1' | 'bleu2' | 'bleu4' | 'rouge1' | 'rouge2' | 'rougel',
  EvaluationValueDelta
>

export type EvaluationCostDeltas = Record<
  'amount' | 'calls' | 'prompt_tokens' | 'completion_tokens' | 'total_tokens' | 'cached_tokens',
  EvaluationValueDelta
>

export type EvaluationTimingDeltas = Record<
  'total_wall_time_ms' | 'preparation_ms' | 'evaluation_ms' | 'cleanup_ms' | 'case_avg_ms' | 'case_p50_ms' | 'case_p95_ms' | 'model_call_cumulative_ms',
  EvaluationValueDelta
>

export interface EvaluationRunComparison {
  run: EvaluationRunSummary
  config: EvaluationRunConfig | null
  quality_compatibility: EvaluationComparisonCompatibility
  cost_compatibility: EvaluationComparisonCompatibility
  timing_compatibility: EvaluationComparisonCompatibility
  quality: EvaluationQualityDeltas
  cost: EvaluationCostDeltas
  timing: EvaluationTimingDeltas
}

export interface EvaluationComparison {
  baseline_id: string
  runs: EvaluationRunComparison[]
}

export interface EvaluationTask {
  id: string
  tenant_id: number
  dataset_id: string
  start_time: string
  status: number
  err_msg?: string
  total: number
  finished: number
}

export interface EvaluationRunResult {
  schema_version: string
  run: {
    run_id: string
    tenant_id: number
    dataset_id: string
    started_at: string
    completed_at: string | null
    status: EvaluationRunStatus
    pricing_version?: string
  }
  retrieval: EvaluationRetrievalResult | null
  answer: EvaluationAnswerResult | null
  usage: EvaluationUsageResult
  cost: EvaluationCostResult
  timing: EvaluationTimingResult
  cases: EvaluationCaseResult[]
  warnings: EvaluationWarning[]
}

export interface EvaluationDetail {
  task: EvaluationTask
  params: Record<string, unknown> | null
  config?: EvaluationRunConfig
  metric?: EvaluationMetrics
  result?: EvaluationRunResult
}

export interface StartEvaluationRequest {
  dataset_id: string
  knowledge_base_id: string
  chat_id: string
  rerank_id?: string
}

export interface EvaluationRunListParams {
  status?: EvaluationRunStatus
  dataset_id?: string
  config_hash?: string
  embedding_model_id?: string
  chat_model_id?: string
  rerank_model_id?: string
  started_from?: string
  started_to?: string
  page?: number
  page_size?: number
}

interface ApiResponse<T> {
  success: boolean
  data: T
}

export function startEvaluation(data: StartEvaluationRequest) {
  return post<ApiResponse<EvaluationDetail>>('/api/v1/evaluation', data)
}

export function getEvaluationTask(taskID: string) {
  return get<ApiResponse<EvaluationDetail>>('/api/v1/evaluation', {
    params: { task_id: taskID },
  })
}

export function listEvaluationRuns(params: EvaluationRunListParams) {
  return get<ApiResponse<EvaluationRunPage>>('/api/v1/evaluation/runs', { params })
}

export function getEvaluationRun(runID: string) {
  return get<ApiResponse<EvaluationRunOverview>>(`/api/v1/evaluation/runs/${encodeURIComponent(runID)}`)
}

export function listEvaluationRunCases(
  runID: string,
  params: { status?: EvaluationRunStatus; page?: number; page_size?: number },
) {
  return get<ApiResponse<EvaluationCasePage>>(
    `/api/v1/evaluation/runs/${encodeURIComponent(runID)}/cases`,
    { params },
  )
}

export function compareEvaluationRuns(baselineID: string, runIDs: string[]) {
  return get<ApiResponse<EvaluationComparison>>('/api/v1/evaluation/comparison', {
    params: {
      baseline_id: baselineID,
      run_ids: runIDs.join(','),
    },
  })
}
